docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,421 @@
+======
+Codecs
+======
+
+.. default-domain:: mongodb
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+.. versionadded:: 1.17
+
+Overview
+--------
+
+Codecs are used to decode BSON documents into PHP objects, and encode PHP objects into BSON documents. In contrast to
+other methods, they allow for more flexibility and customization of the process and how different data types are
+handled. They allow separating the logic for BSON encoding and decoding from the domain classes, allowing to decode BSON
+into plain old PHP objects (POPOs).
+
+Document Codec Usage
+--------------------
+
+The main logic is contained in a document codec. This class implements the ``MongoDB\Codec\DocumentCodec`` interface and
+defines what data types can be encoded/decoded and how. The following example defines a ``Person`` class and a codec to
+encode/decode it:
+
+.. code-block:: php
+
+   <?php
+
+   final class Person
+   {
+       public MongoDB\BSON\ObjectId $id;
+       public string $name;
+
+       public function __construct(string $name)
+       {
+           $this->id = new MongoDB\BSON\ObjectId();
+           $this->name = $name;
+       }
+   }
+
+.. code-block:: php
+
+   <?php
+
+   final class PersonCodec implements MongoDB\Codec\DocumentCodec
+   {
+       // These traits define commonly used functionality to avoid duplication
+       use MongoDB\Codec\DecodeIfSupported;
+       use MongoDB\Codec\EncodeIfSupported;
+
+       public function canDecode($value): bool
+       {
+           return $value instanceof MongoDB\BSON\Document && $value->has('name');
+       }
+
+       public function canEncode($value): bool
+       {
+           return $value instanceof Person;
+       }
+
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $person = new Person($value->get('name'));
+           $person->id = $value->get('_id');
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               '_id' => $value->id,
+               'name' => $value->name,
+           ]);
+       }
+   }
+
+To then use this codec with a collection, specify the ``codec`` option when selecting the collection:
+
+.. code-block:: php
+
+   <?php
+
+   $collection = (new MongoDB\Client())->selectCollection('test', 'person', ['codec' => new PersonCodec()]);
```

Comment:
It's easier to read if things are separate. And you can put the option on its own line to make it stand out.
```suggestion
   $client = new MongoDB\Client();
   $collection = $client->selectCollection('test', 'person', [
       'codec' => new PersonCodec(),
   ]);
```